firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/serializers/TimestampSerializer.kt

```diff
@@ -56,41 +56,50 @@ public object TimestampSerializer : KSerializer<Timestamp> {
     return "$serializedSecond.${serializedNano}Z"
   }
 
-  /**
-   * Note: Timestamp sent from SDK is always 9 digits nanosecond precision, meaning there are 9
-   * digits in SSSSSSSSS parts. However, when running against different databases, this precision
-   * might change, and server will truncate it to 0/3/6 digits precision without throwing an error.
-   */
+  // TODO: Replace this implementation with Instant.parse() once minSdkVersion is bumped to at
+  //  least 26 (Build.VERSION_CODES.O).
   private fun timestampFromString(str: String): Timestamp {
-    // TODO: support +/- time zone offsets as well (b/337299540)
-    require(str.uppercase().endsWith("Z")) {
-      "the last character should be 'Z', but got '${str.takeLast(1)}' (str=$str)"
+    val time = str.uppercase()
+
+    val timeZoneOffsetFormatRegex =
+      Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d{0,9})?Z$")
+    val nanoFormatRegex = Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}[+-]\\d{2}:\\d{2}$")
+
+    require(time.matches(timeZoneOffsetFormatRegex) || time.matches(nanoFormatRegex)) {
+      "str=$time cannot be deserialized into Timestamp."
```

Comment:
Please update this error message to better explain _why_ it cannot be deserialized; for example, you could say something like that the string does not match the RFC3339 format. There is no need to say anything about "deserialization" in the error message since that will be implied by the context.